### PR TITLE
guestagent: watch NETFILTER_CFG audit events to reduce polling

### LINKED ISSUE #258: guestagent: watch netlink events instead of polling
> Probably this ticking is too heavy for watching iptables. Instead we should wait for `NETLINK_AUDIT` events or some eBPF stuff, but I can work on that in a separate PR after merging this.

_Originally posted by @AkihiroSuda in https://github.com/lima-vm/lima/pull/242#discussion_r708812032_

- - -
We can use https://github.com/elastic/go-libaudit

### PR DESCRIPTION
**Pull request by AkihiroSuda**
iptables is scanned only when NETFILTER_CFG audit event was observed in the last minute.
Previously, iptables was scanned every 3 seconds.

`/proc/net/tcp` is still scanned every 3 seconds.

Fix #258

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
cc @jandubois @mattfarina

**Comment by jandubois** [comment]
> iptables is scanned only when NETFILTER_CFG audit event was observed in the last minute.

I don't understand the reasoning behind this. Doesn't it mean it can take up to a minute before a port will be forwarded after you start a container? That feels like an awfully long time.

I'm also not sure about using `tick * 20` for this interval; if you check /proc/net/tcp` every 10 seconds, it means you would wait 3½ minutes before checking iptables.

**Comment by AkihiroSuda** [comment]
When the audit event was observed, the guestagent configures itself to enable “watch iptables every 3 second” mode. (Because watching iptables on every audit event would be too heavy.)
@
When no audit event was observed for 3*20=60 seconds, the guestagent disables “watch iptables every 3 second” mode.

So, port 80 is enabled in about 3 seconds. Almost same delay as port 8080.

File: pkg/guestagent/guestagent_linux.go [hunk 1]
```diff
@@ -5,26 +5,88 @@ import (
...
+// when received NETFILTER_CFG audit message.
+//
+// setWorthCheckingIPTablesRoutine sets worthCheckingIPTables to be false
+// when no NETFILTER_CFG audit message was received for the iptablesIdle time.
+func (a *agent) setWorthCheckingIPTablesRoutine(auditClient *libaudit.AuditClient, iptablesIdle time.Duration) {
+	var latestTrue time.Time
+	go func() {
+		for {
+			time.Sleep(iptablesIdle)
+			a.worthCheckingIPTablesMu.Lock()
```
Comment (jandubois) [inline pkg/guestagent/guestagent_linux.go:65]: Is this whole mutex really needed? Afaict it is just "protecting" a bool and an int, both of which should be updated atomically anyways. And the whole thing is a rough heuristic anyways, so it doesn't even matter if it gets the values wrong in one iteration.

But I would at least like to understand if this is necessary for some reason.

Comment (AkihiroSuda) [reply pkg/guestagent/guestagent_linux.go:65]: `go test -race` will raise errors if we remove the mutex (and we add race tests in future)

**Review by jandubois — COMMENTED** [review]
I'm not really comfortable reviewing this PR because I'm not familiar with the audit framework, and don't know which activities trigger the NETFILTER_CFG events:

1. How often do these events occur? Is there really ever a significant period of time without an event to turn off monitoring of iptables?
2. Will there always be an event whenever a filter rule is modified (I assume this is true from the event name, but don't really know).

I also don't know if access to the bool and time values need to be protected by a mutex, but it obviously won't hurt, so doesn't really matter.

Beyond that, the code looks fine to me.

**Review by jandubois — APPROVED** [review]

File: pkg/guestagent/guestagent_linux.go (hunk 1 above)
Comment (jandubois) [reply pkg/guestagent/guestagent_linux.go:65]: Ok, that's good enough a reason :)

**Comment by AkihiroSuda** [comment]
> How often do these events occur? Is there really ever a significant period of time without an event to turn off monitoring of iptables?

22 events on every `sudo nerdctl run`, 24 events on  every`sudo nerdctl rm`.

> Will there always be an event whenever a filter rule is modified (I assume this is true from the event name, but don't really know).

I assume yes, but not sure it is guaranteed, so we also scan the iptables every minute.

---
Events on `sudo nerdctl run -d --name nginx -p 127.0.0.1:80:80 nginx:alpine`:

```
type=NETFILTER_CFG msg=audit(1633416364.368:230): table=nat:5;CNI-d391c03337f7eb58c39bc9f8:64 family=2 entries=0 op=nft_register_chain pid=74822 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.368:230): table=?:0;?:0 family=0 entries=186 op=nft_register_gen pid=74822 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.372:231): table=nat:5;CNI-d391c03337f7eb58c39bc9f8:64 family=2 entries=65 op=nft_register_rule pid=74824 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.372:231): table=?:0;?:0 family=0 entries=187 op=nft_register_gen pid=74824 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.376:232): table=nat:5;CNI-d391c03337f7eb58c39bc9f8:64 family=2 entries=66 op=nft_register_rule pid=74826 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.376:232): table=?:0;?:0 family=0 entries=188 op=nft_register_gen pid=74826 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.380:233): table=nat:5;POSTROUTING:1 family=2 entries=67 op=nft_register_rule pid=74828 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.380:233): table=?:0;?:0 family=0 entries=189 op=nft_register_gen pid=74828 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.404:234): table=nat:5;CNI-DN-d391c03337f7eb58c39bc:68 family=2 entries=0 op=nft_register_chain pid=74845 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.404:234): table=?:0;?:0 family=0 entries=190 op=nft_register_gen pid=74845 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.408:235): table=nat:5;CNI-DN-d391c03337f7eb58c39bc:68 family=2 entries=69 op=nft_register_rule pid=74847 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.408:235): table=?:0;?:0 family=0 entries=191 op=nft_register_gen pid=74847 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.412:236): table=nat:5;CNI-DN-d391c03337f7eb58c39bc:68 family=2 entries=70 op=nft_register_rule pid=74849 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.412:236): table=?:0;?:0 family=0 entries=192 op=nft_register_gen pid=74849 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.416:237): table=nat:5;CNI-DN-d391c03337f7eb58c39bc:68 family=2 entries=71 op=nft_register_rule pid=74851 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.416:237): table=?:0;?:0 family=0 entries=193 op=nft_register_gen pid=74851 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.416:238): table=nat:5;CNI-HOSTPORT-DNAT:25 family=2 entries=72 op=nft_register_rule pid=74853 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.416:238): table=?:0;?:0 family=0 entries=194 op=nft_register_gen pid=74853 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.440:239): table=filter:3;CNI-FORWARD:40 family=2 entries=59 op=nft_register_rule pid=74867 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.440:239): table=?:0;?:0 family=0 entries=195 op=nft_register_gen pid=74867 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.444:240): table=filter:3;CNI-FORWARD:40 family=2 entries=60 op=nft_register_rule pid=74869 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416364.444:240): table=?:0;?:0 family=0 entries=196 op=nft_register_gen pid=74869 subj=unconfined comm="iptables"
```


Events on `sudo nerdctl rm -f nginx`

```
type=NETFILTER_CFG msg=audit(1633416384.204:249): table=filter:3;CNI-FORWARD:40 family=2 entries=59 op=nft_unregister_rule pid=75047 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.204:249): table=?:0;?:0 family=0 entries=197 op=nft_register_gen pid=75047 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.208:250): table=filter:3;CNI-FORWARD:40 family=2 entries=60 op=nft_unregister_rule pid=75048 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.208:250): table=?:0;?:0 family=0 entries=198 op=nft_register_gen pid=75048 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.216:251): table=nat:5;CNI-DN-d391c03337f7eb58c39bc:68 family=2 entries=69 op=nft_unregister_rule pid=75059 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.216:251): table=nat:5;CNI-DN-d391c03337f7eb58c39bc:68 family=2 entries=70 op=nft_unregister_rule pid=75059 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.216:251): table=nat:5;CNI-DN-d391c03337f7eb58c39bc:68 family=2 entries=71 op=nft_unregister_rule pid=75059 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.216:251): table=?:0;?:0 family=0 entries=199 op=nft_register_gen pid=75059 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.224:252): table=nat:5;CNI-HOSTPORT-DNAT:25 family=2 entries=72 op=nft_unregister_rule pid=75064 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.224:252): table=?:0;?:0 family=0 entries=200 op=nft_register_gen pid=75064 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.228:253): table=nat:5;CNI-DN-d391c03337f7eb58c39bc:68 family=2 entries=0 op=nft_unregister_chain pid=75065 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.228:253): table=?:0;?:0 family=0 entries=201 op=nft_register_gen pid=75065 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.228:254): table=nat:5;CNI-SN-d391c03337f7eb58c39bc:73 family=2 entries=0 op=nft_register_chain pid=75066 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.228:254): table=?:0;?:0 family=0 entries=202 op=nft_register_gen pid=75066 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.232:255): table=nat:5;CNI-SN-d391c03337f7eb58c39bc:73 family=2 entries=0 op=nft_unregister_chain pid=75068 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.232:255): table=?:0;?:0 family=0 entries=203 op=nft_register_gen pid=75068 subj=unconfined comm="iptables"
type=NETFILTER_CFG msg=audit(1633416384.232:256): table=nat:8;CNI-DN-d391c03337f7eb58c39bc:9 family=10 entries=0 op=nft_register_chain pid=75069 subj=unconfined comm="ip6tables"
type=NETFILTER_CFG msg=audit(1633416384.232:256): table=?:0;?:0 family=0 entries=204 op=nft_register_gen pid=75069 subj=unconfined comm="ip6tables"
type=NETFILTER_CFG msg=audit(1633416384.236:257): table=nat:8;CNI-DN-d391c03337f7eb58c39bc:9 family=10 entries=0 op=nft_unregister_chain pid=75071 subj=unconfined comm="ip6tables"
type=NETFILTER_CFG msg=audit(1633416384.236:257): table=?:0;?:0 family=0 entries=205 op=nft_register_gen pid=75071 subj=unconfined comm="ip6tables"
type=NETFILTER_CFG msg=audit(1633416384.240:258): table=nat:8;CNI-SN-d391c03337f7eb58c39bc:10 family=10 entries=0 op=nft_register_chain pid=75072 subj=unconfined comm="ip6tables"
type=NETFILTER_CFG msg=audit(1633416384.240:258): table=?:0;?:0 family=0 entries=206 op=nft_register_gen pid=75072 subj=unconfined comm="ip6tables"
type=NETFILTER_CFG msg=audit(1633416384.240:259): table=nat:8;CNI-SN-d391c03337f7eb58c39bc:10 family=10 entries=0 op=nft_unregister_chain pid=75074 subj=unconfined comm="ip6tables"
type=NETFILTER_CFG msg=audit(1633416384.240:259): table=?:0;?:0 family=0 entries=207 op=nft_register_gen pid=75074 subj=unconfined comm="ip6tables"
```

**Comment by jandubois** [comment]
> > How often do these events occur? Is there really ever a significant period of time without an event to turn off monitoring of iptables?
> 
> 22 events on every `sudo nerdctl run`, 24 events on every`sudo nerdctl rm`.

I was wondering about the opposite thing: if you are not running any commands, is there a quiet period where no events are received, or is there so much background "chatter" from the OS itself that it never really turns off anyways.
 
> > Will there always be an event whenever a filter rule is modified (I assume this is true from the event name, but don't really know).
> 
> I assume yes, but not sure it is guaranteed, so we also scan the iptables every minute.

I don't see this in the code. I see that scanning is turned off when there have been no events for one minute, so if there are no new events, the tables will not be refreshed. I think this will all be fine, given the number of events you showed above; I was simply curious.

**Comment by AkihiroSuda** [comment]
>  is there a quiet period where no events are received

Seems almost always "quiet" when no `nerdctl run/rm` is executed

**Pull request merged**